### PR TITLE
Feature/svc 1111

### PR DESCRIPTION
**Pull request by DeKapito**

### CODE REVIEW

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 1]
```diff
@@ -37,47 +39,91 @@ def config_key(cls, datasource, context):
     def params(cls, datasource, context):
         return {
             'api_key': context.zOWeatherAPIKey,
+            'api_version': context.zOWeatherAPIVersion,
             'city_name': context.city_name,
             'country_code': context.country_code,
             'location_name': context.title,
             }
 
+    def camel_case_to_snake(self, name):
```
Comment (azadorozhnyizenoss) [inline ZenPacks/training/OpenWeather/dsplugins.py]: cover this function by unitests. create module util.py and move it. Docstrings as well.

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 2]
```diff
@@ -37,47 +39,91 @@ def config_key(cls, datasource, context):
...
                     "%s: failed to get conditions data for %s",
                     config.id,
                     datasource.params['location_name'])
 
+                data['events'].append({
+                    'device': config.id,
+                    'component': datasource.component,
+                    'eventKey': 'oweather-conditions',
+                    'eventClassKey': 'oweather',
+                    'severity': SEVERITY_ERROR,
```
Comment (azadorozhnyizenoss) [inline ZenPacks/training/OpenWeather/dsplugins.py]: Clear this event

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 3]
```diff
@@ -37,47 +39,91 @@ def config_key(cls, datasource, context):
...
+                    'summary':
+                        '{}: failed to get conditions data for {}'
+                        .format(config.id, datasource.params['location_name']),
+                    'message':
+                        '{}: failed to get conditions data for {}. Reason: {}'
```
Comment (azadorozhnyizenoss) [inline ZenPacks/training/OpenWeather/dsplugins.py]: no sense to duplicate the same string in summary

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 4]
```diff
@@ -37,47 +39,91 @@ def config_key(cls, datasource, context):
...
                 continue
 
-            weather = response['weather'][0]
-            main = response['main']
-            wind = response['wind']
+            try:
+                weather = response['weather'][0]
+                main = response['main']
+                wind = response['wind']
+            except (KeyError, TypeError):
```
Comment (azadorozhnyizenoss) [inline ZenPacks/training/OpenWeather/dsplugins.py]: IndexError

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 5]
```diff
@@ -37,47 +39,91 @@ def config_key(cls, datasource, context):
...
+                    config.id,
+                    datasource.params['location_name'])
+
+                data['events'].append({
+                    'device': config.id,
+                    'component': datasource.component,
+                    'eventKey': 'oweather-conditions',
+                    'eventClassKey': 'oweather',
+                    'severity': SEVERITY_ERROR,
+                    'message':
```
Comment (azadorozhnyizenoss) [inline ZenPacks/training/OpenWeather/dsplugins.py]: summary=message. Specify summary
Please clear event
Generate different event for specific case

File: ZenPacks/training/OpenWeather/modeler/plugins/OpenWeather/Locations.py [hunk 6]
```diff
@@ -21,61 +23,117 @@ class Locations(PythonPlugin):
...
     def collect(self, device, log):
         """Asynchronously collect data from device. Return a deferred."""
         log.info("%s: collecting data", device.id)
+        self._eventService = getUtility(IEventService)
 
         apikey = getattr(device, 'zOWeatherAPIKey', None)
         if not apikey:
-            log.error(
-                "%s: %s not set. Get one from https://openweathermap.org/api",
+            message = '{}: {} not set. Get one from https://openweathermap.org/api'\
```
Comment (azadorozhnyizenoss) [inline ZenPacks/training/OpenWeather/modeler/plugins/OpenWeather/Locations.py]: refactor this please :)

File: ZenPacks/training/OpenWeather/modeler/plugins/OpenWeather/Locations.py [hunk 7]
```diff
@@ -21,61 +23,117 @@ class Locations(PythonPlugin):
...
+                for result in response['RESULTS']:
+                    rm.append(self.objectMap({
+                        'id': self.prepId(result['zmw']),
+                        'title': result['name'],
+                        'api_link': result['l'],
+                        'city_name': result['name'].split(',')[0],
+                        'country_code': result['c'],
+                        'timezone': result['tzs'],
+                        }))
+            except (KeyError, TypeError), e:
```
Comment (azadorozhnyizenoss) [inline ZenPacks/training/OpenWeather/modeler/plugins/OpenWeather/Locations.py]: add ValueError for prepId, and IndexError

File: ZenPacks/training/OpenWeather/zenpack.yaml [hunk 8]
```diff
@@ -58,31 +61,39 @@ device_classes:
...
+              tempMin: GAUGE
+              tempMax: GAUGE
               speed: GAUGE
 
+        thresholds:
+          "Low temperature":
+            dsnames: [conditions_temp]
+            type: MinMaxThreshold
+            minval: -2.0
+            eventClass: /Status
```
Comment (azadorozhnyizenoss) [inline ZenPacks/training/OpenWeather/zenpack.yaml]: event with eventClass /Status is used for determine device status (UP or Down)

**Review by azadorozhnyizenoss — CHANGES_REQUESTED** [review]
👍 +1

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 9]
```diff
@@ -65,78 +62,86 @@ def collect(self, config):
...
+                    'eventClassKey': 'oweatherConditions',
                     'severity': SEVERITY_ERROR,
-                    'summary':
-                        '{}: failed to get conditions data for {}'
-                        .format(config.id, datasource.params['location_name']),
-                    'message':
-                        '{}: failed to get conditions data for {}. Reason: {}'
-                        .format(config.id, datasource.params['location_name'], e.message),
+                    'summary': message,
+                    'message': e,
```
Comment (ayavorskyi) [inline ZenPacks/training/OpenWeather/dsplugins.py]: make summary more concise and send message as `message`

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 10]
```diff
@@ -65,78 +62,86 @@ def collect(self, config):
...
                 })
                 continue
+            else:
+                data['events'].append({
+                    'device': config.id,
+                    'component': datasource.component,
+                    'eventKey': 'oweatherConditionsCollect_%s' % datasource.params['location_name'],
+                    'eventClassKey': 'oweatherConditions',
+                    'severity': SEVERITY_CLEAR,
+                    'summary': '',
```
Comment (ayavorskyi) [inline ZenPacks/training/OpenWeather/dsplugins.py]: add some message

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 11]
```diff
@@ -65,78 +62,86 @@ def collect(self, config):
...
-                        .format(config.id, datasource.params['location_name'], e.message),
+                    'summary': message,
+                    'message': e,
                 })
                 continue
+            else:
+                data['events'].append({
+                    'device': config.id,
+                    'component': datasource.component,
+                    'eventKey': 'oweatherConditionsCollect_%s' % datasource.params['location_name'],
```
Comment (ayavorskyi) [inline ZenPacks/training/OpenWeather/dsplugins.py]: feel free to use format() :)

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 12]
```diff
@@ -65,78 +62,86 @@ def collect(self, config):
 
                 response = json.loads(response)
             except Exception, e:
```
Comment (ayavorskyi) [inline ZenPacks/training/OpenWeather/dsplugins.py]: replace , with as

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 13]
```diff
@@ -65,78 +62,86 @@ def collect(self, config):
...
-                    'message':
-                        '{}: failed to get conditions data for {}'
-                            .format(config.id, datasource.params['location_name']),
+                    'summary': message,
+                    'message': e
                 })
 
                 returnValue(data)
+            else:
+                data['events'].append({
```
Comment (ayavorskyi) [inline ZenPacks/training/OpenWeather/dsplugins.py]: think about creating a generic method for event creation

File: ZenPacks/training/OpenWeather/dsplugins.py [hunk 14]
```diff
@@ -65,78 +62,86 @@ def collect(self, config):
...
 
                     except (TypeError, ValueError):
                         # Sometimes values are NA or not available.
                         continue
 
-                    dpname = '_'.join((datasource.datasource, datapoint_id))
+                    dpname = '_'.join((datasource.datasource, datapointId))
                     data['values'][datasource.component][dpname] = (value, 'N')
 
             data['maps'].append(
```
Comment (ayavorskyi) [inline ZenPacks/training/OpenWeather/dsplugins.py]: keep in mind in real world it's not the best idea to follow :) 
take a look at datasource property of components fileds
https://zenpack-sdk.zenoss.com/en/latest/yaml-classes-and-relationships.html#class-property-fields

File: ZenPacks/training/OpenWeather/modeler/plugins/OpenWeather/Locations.py [hunk 15]
```diff
@@ -29,99 +29,99 @@ class Locations(PythonPlugin):
...
-                device.id,
-                SEVERITY_ERROR,
-                message
-            )
+        apikey = self.getProperty(device, 'zOWeatherAPIKey', log)
+        locations = self.getProperty(device, 'zOWeatherLocations', log)
+        apiver = self.getProperty(device, 'zOWeatherAPIVersion', log)
 
+        # Check apikey, apiver, locations. If one is missing then return None
+        if not (apikey and apiver and locations):
```
Comment (ayavorskyi) [inline ZenPacks/training/OpenWeather/modeler/plugins/OpenWeather/Locations.py]: take a look at any/all in python

File: ZenPacks/training/OpenWeather/modeler/plugins/OpenWeather/Locations.py [hunk 16]
```diff
@@ -29,99 +29,99 @@ class Locations(PythonPlugin):
...
 
     def process(self, device, results, log):
         """Process results. Return iterable of datamaps or None."""
 
         rm = self.relMap()
 
-        for response in results:
+        for location, response in results:
             try:
+                if len(response['RESULTS']) == 0:
```
Comment (ayavorskyi) [inline ZenPacks/training/OpenWeather/modeler/plugins/OpenWeather/Locations.py]: if response['RESULTS']

**Review by ayavorskyi — CHANGES_REQUESTED** [review]
good job 
keep on 👍

**Review by azadorozhnyizenoss — APPROVED** [review]
well done!

**Pull request merged**